SQL: Fix catalog filtering in SYS COLUMNS

costin opened the pull request:
Properly treat '%' as a wildcard for catalog filtering instead of doing
a straight string match.

Fix #40582

costin committed:
SQL: Fix catalog filtering in SYS COLUMNS

Properly treat '%' as a wildcard for catalog filtering instead of doing
a straight string match.

Fix #40582

elasticmachine commented:
Pinging @elastic/es-search

costin committed:
Table filtering now considers aliases as well.
Add integration test for SYS COLUMNS - currently running only inside
single_node since the cluster name is test dependent.

costin commented:
I have updated the PR to fix also alias filtering which was not performed (only the index name was considered). Also added a test suite since mocking doesn't really work for this case.

matriv commented on x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/JdbcCsvSpecIT.java:
```diff
@@ -5,10 +5,26 @@
+    public static List<Object[]> readScriptSpec() throws Exception {
+        List<Object[]> list = new ArrayList<>();
+        list.addAll(CsvSpecTestCase.readScriptSpec());
+        return readScriptSpec("/disabled/command-sys.csv-spec", specParser());
```
I don't like `disabled` in the path here.

astefan reviewed (approved):
LGTM

costin committed:
Address feedback

matriv reviewed (approved):
LGTM

costin committed:
Add escaping char for LIKE queries with user defined params

costin committed:
Ignore nested fields always (since SYS TABLE is a meta command meant
for the drivers)
Move away from separateMappings() since in case of aliases pointing to
multiple indices, it can return duplicated entries (as there's no
merging) causing issues.
Hence why mergedMapping is used instead with the assumption that only
one table will be used at a time.
A more correct (yet quite expensive) alternative would be to return the
mapping per index yet in case of aliases, drop it and use field caps
instead. However as this is not CCS-friendly, using field_caps instead
seems a good solution for now.

costin committed:
update tests

costin committed:
Keep using separate mapping for getColumn '%'
Fix monotony of ORDINAL_POSITION

costin committed:
Merge branch 'master' into fix-40582

matriv reviewed (approved):
LGTM

astefan commented on x-pack/plugin/sql/qa/src/main/resources/single-node-only/command-sys.csv-spec:
```diff
@@ -0,0 +1,120 @@
+sysColumnsWithTableLikeNoEscape
+SYS COLUMNS TABLE LIKE 'test_emp';
+
+// since there's no escaping test_emp means test*emp which matches also test_alias_emp
```
`test_emp` should match `test`[one character]`emp` when there is no escape character, no?

costin replied:
Unfortunately due to the current field caps features, we can't figure out the source index.

astefan reviewed (commented):
Added two comments, but am not 100% sure I am right about those.

bpintea commented:
Gave the branch a try and if nothing was mistaken during the process:
* `sys columns catalog '%' table like 'employees'` (and `sys columns catalog '%' table like 'employees' like '%'`) returns an empty result set;
* `sys columns catalog 'distribution\_run'  table like '%'` will return a full list of all columns in all tables is the catalog (which is right); but `sys columns catalog 'distribution\_run'  table like ''` will return the same, except for the table name column which is empty (which seems wrong).

costin committed:
Merge remote-tracking branch 'remotes/upstream/master' into fix-40582

costin commented on x-pack/plugin/sql/qa/src/main/resources/single-node-only/command-sys.csv-spec:
```diff
@@ -0,0 +1,120 @@
+
+   TABLE_CAT:s |  TABLE_SCHEM:s|  TABLE_NAME:s |    COLUMN_NAME:s   |   DATA_TYPE:i |   TYPE_NAME:s |  COLUMN_SIZE:i| BUFFER_LENGTH:i|DECIMAL_DIGITS:i|NUM_PREC_RADIX:i | NULLABLE:i|    REMARKS:s  |  COLUMN_DEF:s |SQL_DATA_TYPE:i|SQL_DATETIME_SUB:i|CHAR_OCTET_LENGTH:i|ORDINAL_POSITION:i|IS_NULLABLE:s|SCOPE_CATALOG:s|SCOPE_SCHEMA:s|SCOPE_TABLE:s|SOURCE_DATA_TYPE:sh|IS_AUTOINCREMENT:s|IS_GENERATEDCO ...
+---------------+---------------+---------------+--------------------+---------------+---------------+---------------+---------------+---------------+---------------+---------------+---------------+---------------+---------------+----------------+-----------------+----------------+---------------+---------------+---------------+---------------+----------------+----------------+------------------
+x-pack_plugin_sql_qa_single-node_integTestCluster  |null           |test\_emp      |birth_date        |93             |DATETIME       |24             |8              |null           |null           |1              |null           |null           |9              |3               |null             |1               |YES            |null           |null           |null           |null            |NO  ...
```
This is a bug - the `test\_emp` is escaped so the name should be `test_emp`.

costin committed:
Try pattern unescaping

bpintea commented:
Update after consulting with @costin:

> `sys columns catalog '%' table like 'employees'` (and `sys columns catalog '%' table like 'employees' like '%'`) returns an empty result set;

This (=empty result set) is correct, since the catalog argument [`cannot contain a string search pattern`](https://docs.microsoft.com/en-us/sql/odbc/reference/syntax/sqlcolumns-function).

> `sys columns catalog 'distribution\_run' table like ''` will return the same, except for the table name column which is empty (which seems wrong)

This is currently a known limitation; however, it should have no practical impact, since the applications ask for the columns of specific table.
